Add mask operation on IPv4 and IPv6 interfaces

dominik-lekse opened the pull request:
This pull requests adds the `mask` operation to `math`. The mask operation applies an integer network mask to a given input IPv4 or IPv6 address.  

### Documentation

The functionality is described in the added documentation as follows.  

`mask`: Applies the given network mask to the address. The network mask is
  	expressed as a decimal value (e.g. network mask "24" corresponds to
  	`255.255.255.0`). After applying the network mask, the network mask of the
  	resulting address will be either the applied network mask or the network mask
  	of the input address depending on which network is larger
  	(e.g. 192.168.10.20/24 `"mask" "16"` will return "192.168.0.0/16" but
  	192.168.10.20/24 `"mask" "28"` will return "192.168.10.16/24").

### Use Cases

Simple example:

```
{{ GetPrivateInterfaces | include "type" "IP" | math "mask" "24" | attr "address" }}
```

I use this operation in a use case in which I need to calculate an IP address relative to a subnet which is larger than the subnet associated to the input interface. In this case, I could not use the `network` operation which is limited to the subnet mask of the input interface.

```
{{ GetPrivateInterfaces | include "type" "IPv4" | math "mask" "20" | math "network" "+4" | attr "address" }}
```

magiconair commented:
@dominik-lekse I just looked at https://github.com/hashicorp/consul/pull/2932 which you're also favoring. Would this PR allow implementing this as well?

sean- commented on ifaddrs.go:
```diff
@@ -866,6 +867,80 @@ func IfAddrMath(operation, value string, inputIfAddr IfAddr) (IfAddr, error) {
+				Interface: inputIfAddr.Interface,
+			}, nil
+		case TypeIPv6:
+			i, err := strconv.ParseInt(value, 10, 32)
```
Can you change this to `ParseUint` instead?

sean- commented on ifaddr_test.go:
```diff
@@ -411,6 +411,102 @@ func TestIfAddrMath(t *testing.T) {
+			},
+			operation: "mask",
+			value:     "16",
+			expected:  "192.168.0.0/16",
```
Thank you for testing a non-zero value (i.e. using `192.168.10.20` vs `192.168.0.0`).  Can you fuzz the upper bound, too, at `.255`?

sean- commented on ifaddrs.go:
```diff
@@ -866,6 +867,80 @@ func IfAddrMath(operation, value string, inputIfAddr IfAddr) (IfAddr, error) {
+				return IfAddr{}, fmt.Errorf("unable to convert %q to int for operation %q: %v", value, operation, err)
+			}
+
+			if i < 0 || i > 32 {
```
We don't need the `i < 0` check since we did a `ParseUint()`.

sean- commented on ifaddrs.go:
```diff
@@ -866,6 +867,80 @@ func IfAddrMath(operation, value string, inputIfAddr IfAddr) (IfAddr, error) {
+				return IfAddr{}, fmt.Errorf("unable to convert %q to int for operation %q: %v", value, operation, err)
+			}
+
+			if i < 0 || i > 128 {
```
Same here, no `i < 0` required after a `ParseUint()`

sean- commented on template/doc.go:
```diff
@@ -212,13 +212,21 @@ Supported operations include:
+  	resulting address will be either the applied network mask or the network mask
+  	of the input address depending on which network is larger
+  	(e.g. 192.168.10.20/24 `"mask" "16"` will return "192.168.0.0/16" but
+  	192.168.10.20/24 `"mask" "28"` will return "192.168.10.16/24").
```
Thank you for that example.

dominik-lekse commented:
@sean- Thanks for the review. In the recent commit, I have addressed your comment.

Regarding the additional tests on the upper bound, I assumed you referred to the host part of the input ipv4 addresses.

Looking forward for the merge.

magiconair commented:
LGTM.
